src/project/manifest/python.rs

```diff
@@ -124,14 +136,70 @@ impl<'de> Deserialize<'de> for VersionOrStar {
     }
 }
 
+#[derive(Debug, Serialize, Clone, PartialEq, Eq, Hash)]
+#[serde(untagged, rename_all = "snake_case", deny_unknown_fields)]
+pub enum GitRev {
+    Short(String),
+    Full(String),
+}
+
+impl GitRev {
+    fn as_full(&self) -> Option<&str> {
+        match self {
+            GitRev::Full(full) => Some(full.as_str()),
+            GitRev::Short(_) => None,
+        }
+    }
+
+    fn to_git_reference(&self) -> GitReference {
+        match self {
+            GitRev::Full(rev) => GitReference::FullCommit(rev.clone()),
+            GitRev::Short(rev) => GitReference::BranchOrTagOrCommit(rev.clone()),
+        }
+    }
+}
+
+impl From<&str> for GitRev {
+    fn from(s: &str) -> Self {
+        if s.len() == 40 {
+            GitRev::Full(s.to_string())
+        } else {
+            GitRev::Short(s.to_string())
+        }
+    }
+}
+
+impl ToString for GitRev {
+    fn to_string(&self) -> String {
+        match self {
+            GitRev::Short(s) => s.clone(),
+            GitRev::Full(s) => s.clone(),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for GitRev {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s: String = Deserialize::deserialize(deserializer)?;
+        if s.len() == 40 {
+            Ok(GitRev::Full(s))
+        } else {
+            Ok(GitRev::Short(s))
+        }
+    }
+}
```

Comment:
Maybe this deserves its own module?

Reply:
I would not move it in separate module as it coupled with PyPiRequirement, which makes sense to keep in `manifest/python.rs`

Reply:
could also go into a submodule of python.rs?

Reply:
I actually moved everything related in a pypi module, have a look!